Awesome stuff there, good job! 👌 

I have tested a couple of basic scenarios, in which 3 of them work properly but 1 (unfortunately) fails:
- ### Modifying components returned by skin transformer
    Works correctly now, tested by the regular `TestSceneFruitObjects (catch) -> LegacyFruitPiece`, also tested with the scenario I got this issue on at first `TestSceneComboCounter (catch) -> LegacyComboCounter` (from #9769)

    Therefore this PR can mark https://github.com/ppy/osu-framework/issues/3775 as resolved!

- ### Modifying drawable hit objects in a player test scene
    Tested with: `TestSceneReplay -> DrawableHitCircle (or any other object)`
    Works correctly.

- ### Modifying mod code in a player test scene
    Tested with: `TestSceneOsuModDoubleTime (osu) -> ModDoubleTime`
    Works correctly.

- ### Modifying the HUD overlay or one of its components in a player test scene
    Tested with: `TestSceneReplay -> HUDOverlay (or any of its children)`
     Fails unfortunately with the following log at `runtime`:
     ```
    [runtime] 2020-09-04 12:26:41 [error]: Error with dynamic compilation!
    [runtime] 2020-09-04 12:26:41 [error]: System.AggregateException: One or more errors occurred. (/Users/salman/Desktop/osu/osu.Game/Rulesets/Mods/ModCinema.cs(25,43): error CS0535: 'ModCinema' does not implement interface member 'IApplicableToHUD.ApplyToHUD(HUDOverlay)') (/Users/salman/Desktop/osu/osu.Game/Tests/Visual/TestPlayer.cs(29,45): error CS0029: Cannot implicitly convert type 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game.Tests/bin/Debug/netcoreapp3.1/osu.Game.dll]' to 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game/Screens/Play/HUDOverlay.cs(25)]') (/Users/salman/Desktop/osu/osu.Game.Rulesets.Catch/Mods/CatchModCinema.cs(18,26): error CS0122: 'CatchAutoGenerator' is inaccessible due to its protection level) (/Users/salman/Desktop/osu/osu.Game.Tests/Visual/Gameplay/TestSceneReplay.cs(35,49): error CS0029: Cannot implicitly convert type 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game.Tests/bin/Debug/netcoreapp3.1/osu.Game.dll]' to 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game/Screens/Play/HUDOverlay.cs(25)]')
    [runtime] 2020-09-04 12:26:41 [error]: ---> System.InvalidOperationException: /Users/salman/Desktop/osu/osu.Game/Rulesets/Mods/ModCinema.cs(25,43): error CS0535: 'ModCinema' does not implement interface member 'IApplicableToHUD.ApplyToHUD(HUDOverlay)'
    [runtime] 2020-09-04 12:26:41 [error]: --- End of inner exception stack trace ---
    [runtime] 2020-09-04 12:26:41 [error]: at osu.Framework.Testing.DynamicClassCompiler`1.recompileAsync(Type targetType, String changedFile) in /Users/salman/Desktop/osu-framework/osu.Framework/Testing/DynamicClassCompiler.cs:line 196
    [runtime] 2020-09-04 12:26:41 [error]: ---> (Inner Exception #1) System.InvalidOperationException: /Users/salman/Desktop/osu/osu.Game/Tests/Visual/TestPlayer.cs(29,45): error CS0029: Cannot implicitly convert type 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game.Tests/bin/Debug/netcoreapp3.1/osu.Game.dll]' to 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game/Screens/Play/HUDOverlay.cs(25)]'<---
    [runtime] 2020-09-04 12:26:41 [error]: 
    [runtime] 2020-09-04 12:26:41 [error]: ---> (Inner Exception #2) System.InvalidOperationException: /Users/salman/Desktop/osu/osu.Game.Rulesets.Catch/Mods/CatchModCinema.cs(18,26): error CS0122: 'CatchAutoGenerator' is inaccessible due to its protection level<---
    [runtime] 2020-09-04 12:26:41 [error]: 
    [runtime] 2020-09-04 12:26:41 [error]: ---> (Inner Exception #3) System.InvalidOperationException: /Users/salman/Desktop/osu/osu.Game.Tests/Visual/Gameplay/TestSceneReplay.cs(35,49): error CS0029: Cannot implicitly convert type 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game.Tests/bin/Debug/netcoreapp3.1/osu.Game.dll]' to 'osu.Game.Screens.Play.HUDOverlay [/Users/salman/Desktop/osu/osu.Game/Screens/Play/HUDOverlay.cs(25)]'<---
    [runtime] 2020-09-04 12:26:41 [error]:
    ```

    - [dynamic compilation log](https://github.com/ppy/osu/files/5175081/dynamic-compilation.log)

--- a/osu.Android.props
+++ b/osu.Android.props
@@ -52,6 +52,6 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="ppy.osu.Game.Resources" Version="2020.904.0" />
-    <PackageReference Include="ppy.osu.Framework.Android" Version="2020.904.0" />
+    <PackageReference Include="ppy.osu.Framework.Android" Version="2020.907.0" />
   </ItemGroup>
 </Project>

--- a/osu.Game/osu.Game.csproj
+++ b/osu.Game/osu.Game.csproj
@@ -24,7 +24,7 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="2.2.6" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="2.2.6" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
-    <PackageReference Include="ppy.osu.Framework" Version="2020.904.0" />
+    <PackageReference Include="ppy.osu.Framework" Version="2020.907.0" />
     <PackageReference Include="ppy.osu.Game.Resources" Version="2020.904.0" />
     <PackageReference Include="Sentry" Version="2.1.6" />
     <PackageReference Include="SharpCompress" Version="0.26.0" />

--- a/osu.iOS.props
+++ b/osu.iOS.props
@@ -70,7 +70,7 @@
     <Reference Include="System.Net.Http" />
   </ItemGroup>
   <ItemGroup Label="Package References">
-    <PackageReference Include="ppy.osu.Framework.iOS" Version="2020.904.0" />
+    <PackageReference Include="ppy.osu.Framework.iOS" Version="2020.907.0" />
     <PackageReference Include="ppy.osu.Game.Resources" Version="2020.904.0" />
   </ItemGroup>
   <!-- Xamarin.iOS does not automatically handle transitive dependencies from NuGet packages. -->
@@ -80,7 +80,7 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="2.2.6" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="2.2.6" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
-    <PackageReference Include="ppy.osu.Framework" Version="2020.904.0" />
+    <PackageReference Include="ppy.osu.Framework" Version="2020.907.0" />
     <PackageReference Include="SharpCompress" Version="0.26.0" />
     <PackageReference Include="NUnit" Version="3.12.0" />
     <PackageReference Include="SharpRaven" Version="2.4.0" />

